### PR TITLE
feat(lambda-python): support poetry packaging for PythonFunction

### LINKED ISSUE #11753: (aws-lambda-python): Manage dependencies with Poetry
Are there any plans to support dependency management using Poetry in the same way that requirements.txt and Pipfile are?

Use Case

Poetry is a increasingly popular dependency management/packaging solution for Python and is arguably more user friendly than either pipenv or pip. It would be great if this method had direct support.

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by setu4993**
This PR adds [poetry](https://python-poetry.org/) support to building Lambda dependencies for the PythonFunction.

I used the pipenv setup for inspiration. I'm not very familiar with writing integration tests, so there's a possibility that I could've missed something.

Closes #11753.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-lambda-python/test/bundling.test.ts [hunk 1]
```diff
@@ -140,6 +140,20 @@ describe('Dependency detection', () => {
...
+      if (
+        /pyproject.toml/.test(p.toString()) &&
+        /poetry.lock/.test(p.toString())
+      ) {
+        return true;
+      }
+      return existsSyncOriginal(p);
+    });
+
+    expect(hasDependencies('/asset-input')).toEqual(true);
```
Comment (setu4993) [inline packages/@aws-cdk/aws-lambda-python/test/bundling.test.ts:154]: Test is failing here. Not sure why 🤔. Maybe my `if` condition in the mock is incorrect? The input path is expected to have both `pyproject.toml` and `poetry.lock` files.

File: packages/@aws-cdk/aws-lambda-python/test/integ.function.poetry.py27.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+/*
+ * Stack verification steps:
+ * aws lambda invoke --function-name <function name> --invocation-type Event --payload $(base64 <<<'"OK"') response.json
+ */
+
+class TestStack extends Stack {
+  constructor(scope: Construct, id: string, props?: StackProps) {
+    super(scope, id, props);
+
+    const fn = new lambda.PythonFunction(this, 'my_handler', {
```
Comment (eladb) [inline packages/@aws-cdk/aws-lambda-python/test/integ.function.poetry.py27.ts:17]: Can you please consolidate all these integration tests into a single test with multiple `PythonFunction`s?

**Review by eladb — CHANGES_REQUESTED** [review]
Looks solid. Minor request about integ tests (would be great if you could consolidate the other tests as well).

Please refer to the [contribution guide](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md#integration-tests) for more info about integ tests.

File: packages/@aws-cdk/aws-lambda-python/test/integ.function.poetry.py27.ts (hunk 2 above)
Comment (setu4993) [reply packages/@aws-cdk/aws-lambda-python/test/integ.function.poetry.py27.ts]: Done!

**Comment by setu4993** [comment]
@eladb : Replaced the multiple files for integration tests with a single file that has multiple functions.

Happy to consolidate the various `pipenv` tests into a single file in a separate PR in the next few days, since I do feel that'd be outside the scope of this PR.

**Review by eladb — APPROVED** [review]

**Comment by setu4993** [comment]
Also wanted to mention that I couldn't actually run `yarn integ` in GitPod which kept complaining about unavailability of Docker socket (Docker-in-Docker doesn't appear to be supported in GitPod), so wasn't sure how to test it except with the CodeBuild CI.

**Comment by setu4993** [comment]
Thank you, @eladb!

**Pull request merged**